compiler/qsc_parse/src/lex/formatter.rs

```diff
@@ -0,0 +1,270 @@
+                // Remove any whitespace at the start of a file
+                if three.span.lo != 0 {
+                    vec![Edit::new(0, three.span.lo, "")]
+                } else {
+                    vec![]
+                }
+            }
+            _ => {
+                // not enough tokens to apply a rule
+                // TODO: we'll probably need to handle end-of-file cases here
+                continue;
+            }
+        };
+
+        edits.append(&mut edits_for_triple);
+    }
+
+    edits
+}
+
+fn fix_whitespace(whitespace: &str, indent_level: usize) -> String {
+    //
+    // when you see newline, insert the indent string
+    // and trim until the next newline or the end of the string
+    //
+
+    let mut count_newlines = whitespace.chars().filter(|c| *c == '\n').count();
+
+    // There should always be at least one newline
+    if count_newlines < 1 {
+        count_newlines = 1;
+    }
+    let mut new = "\n".repeat(count_newlines);
+    new.push_str(&make_indent_string(indent_level));
+    new
+}
+
+fn apply_rules(
+    left: &ConcreteToken,
+    whitespace: &str,
+    right: &ConcreteToken,
+    code: &str,
+    indent_level: usize,
+) -> Vec<Edit> {
+    let mut edits = vec![];
+    // when we get here, neither left nor right should be whitespace
+
+    // if the right is a close brace, the indent level should be one less
+    let indent_level = if let ConcreteTokenKind::Cooked(TokenKind::Close(Delim::Brace)) = right.kind
+    {
+        if indent_level > 0 {
+            indent_level - 1
+        } else {
+            indent_level
+        }
+    } else {
+        indent_level
+    };
+
+    match (&left.kind, &right.kind) {
+        (
+            ConcreteTokenKind::Cooked(TokenKind::Open(l)),
+            ConcreteTokenKind::Cooked(TokenKind::Close(r)),
+        ) if l == r => {
+            rule_no_space(left, whitespace, right, &mut edits);
+        }
+        (ConcreteTokenKind::Comment | ConcreteTokenKind::Cooked(TokenKind::DocComment), _) => {
+            rule_trim_comments(left, &mut edits, code);
+            rule_indentation(left, whitespace, right, &mut edits, indent_level);
+        }
+        (ConcreteTokenKind::Cooked(TokenKind::Semi), _) => match &right.kind {
+            ConcreteTokenKind::Comment => {
+                if whitespace.contains('\n') {
+                    rule_indentation(left, whitespace, right, &mut edits, indent_level);
+                }
+            }
+            _ => {
+                rule_indentation(left, whitespace, right, &mut edits, indent_level);
+            }
+        },
+        (_, ConcreteTokenKind::Cooked(TokenKind::Close(Delim::Brace))) => {
+            rule_indentation(left, whitespace, right, &mut edits, indent_level);
+        }
+        (
+            _,
+            ConcreteTokenKind::Cooked(TokenKind::Keyword(
+                Keyword::Operation | Keyword::Function | Keyword::Newtype | Keyword::Namespace,
```

Comment:
I know your version is technically shorter, but I wonder if this would look better if you expand this out to be 4 separate lines like 

```rs
 ConcreteTokenKind::Cooked(TokenKind::Keyword(Keyword::Operation)) |
 ConcreteTokenKind::Cooked(TokenKind::Keyword(Keyword::Function)) |
 ConcreteTokenKind::Cooked(TokenKind::Keyword(Keyword::Newtype)) |
 ConcreteTokenKind::Cooked(TokenKind::Keyword(Keyword::Namespace)) |
```

I'm hoping that if we can have everything lined up, looking like a table, things will be much easier to follow. Sometimes repetition is actually better IMO!

Reply:
done